### PR TITLE
Replace pinned GCHandles with unsafe code.

### PR DESCRIPTION
**Pull request by rolfbjarne**
Creating/freeing GCHandles is much slower than using unsafe code,
so just avoid it when we can.

### CODE REVIEW

File: src/AudioUnit/AudioUnit.cs [hunk 1]
```diff
@@ -1106,13 +1106,8 @@ public AudioUnitStatus SetScheduledFiles (AudioFile[] audioFiles)
...
 				handles [i] = audioFiles [i].Handle;
 
-			GCHandle gch = GCHandle.Alloc (handles, GCHandleType.Pinned);
-			IntPtr ptr = gch.AddrOfPinnedObject ();
-
-			var status = AudioUnitSetProperty (Handle, AudioUnitPropertyIDType.ScheduledFileIDs, AudioUnitScopeType.Global, 0, ptr,  Marshal.SizeOf (ptr) * count);
-
-			gch.Free ();
-			return status;
+			fixed (IntPtr* ptr = &handles [0])
```
Comment (spouliot) [inline src/AudioUnit/AudioUnit.cs]: I think this pattern results in an IndexOutOfRangeException is the method input is `new AudioFile[0]` as there won't be an item `0`. OTOH you can take the address of `handles` itself

File: src/CoreFoundation/CFDataBuffer.cs [hunk 2]
```diff
@@ -36,17 +36,16 @@ class CFDataBuffer : IDisposable {
...
 			this.buffer = buffer;
 
 			/*
 			 * Copy the buffer to allow the native side to take ownership.
 			 */
 
-			var gch = GCHandle.Alloc (buffer, GCHandleType.Pinned);
-			data = CFData.FromData (gch.AddrOfPinnedObject (), buffer.Length);
-			gch.Free ();
+			fixed (byte* ptr = &buffer [0])
```
Comment (spouliot) [inline src/CoreFoundation/CFDataBuffer.cs]: same, `new byte[0]` as input would fail

File: src/CoreFoundation/CFReadStream.cs [hunk 3]
```diff
@@ -151,14 +151,8 @@ public nint Read (byte[] buffer, int offset, int count)
...
 				throw new ArgumentException ();
-			var gch = GCHandle.Alloc (buffer, GCHandleType.Pinned);
-			try {
-				var addr = gch.AddrOfPinnedObject ();
-				var ptr = new IntPtr (addr.ToInt64 () + offset);
-				return CFReadStreamRead (Handle, ptr, count);
-			} finally {
-				gch.Free ();
-			}
+			fixed (byte* ptr = &buffer [0])
```
Comment (spouliot) [inline src/CoreFoundation/CFReadStream.cs]: same

File: src/CoreFoundation/CFWriteStream.cs [hunk 4]
```diff
@@ -111,14 +111,8 @@ public int Write (byte[] buffer, nint offset, nint count)
...
 				throw new ArgumentException ();
-			var gch = GCHandle.Alloc (buffer, GCHandleType.Pinned);
-			try {
-				var addr = gch.AddrOfPinnedObject ();
-				var ptr = new IntPtr (addr.ToInt64 () + offset);
-				return (int) CFWriteStreamWrite (Handle, ptr, count);
-			} finally {
-				gch.Free ();
-			}
+			fixed (byte* ptr = &buffer [0])
```
Comment (spouliot) [inline src/CoreFoundation/CFWriteStream.cs]: same

File: src/CoreMedia/CMBlockBuffer.cs [hunk 5]
```diff
@@ -184,18 +184,17 @@ public CMBlockBufferError CopyDataBytes (nuint offsetToData, nuint dataLength, I
...
 		{
 			if (Handle == IntPtr.Zero)
 				throw new ObjectDisposedException ("BlockBuffer");
 
 			destination = new byte [dataLength];
-			GCHandle destPinned = GCHandle.Alloc (destination, GCHandleType.Pinned);
-			IntPtr destPtr = destPinned.AddrOfPinnedObject ();
-			var error = CMBlockBufferCopyDataBytes (handle, offsetToData, dataLength, destPtr);
+			CMBlockBufferError error;
+			fixed (byte* ptr = &destination [0])
```
Comment (spouliot) [inline src/CoreMedia/CMBlockBuffer.cs]: same, `dataLength` could be `0`

File: src/CoreMedia/CMBlockBuffer.cs [hunk 6]
```diff
@@ -214,19 +213,15 @@ public CMBlockBufferError ReplaceDataBytes (IntPtr sourceBytes, nuint offsetInto
...
 			if (sourceBytes == null)
 				throw new ArgumentNullException (nameof (sourceBytes));
 
-			GCHandle replacePinned = GCHandle.Alloc (sourceBytes, GCHandleType.Pinned);
-			IntPtr replacePtr = replacePinned.AddrOfPinnedObject ();
-
-			var error = ReplaceDataBytes (replacePtr, offsetIntoDestination, (nuint) sourceBytes.Length);
-			replacePinned.Free ();
-			return error;
+			fixed (byte* ptr = &sourceBytes [0])
```
Comment (spouliot) [inline src/CoreMedia/CMBlockBuffer.cs]: same

File: src/SpriteKit/SKWarpGeometryGrid.cs [hunk 7]
```diff
@@ -29,30 +29,13 @@ public static SKWarpGeometryGrid Create (nint cols, nint rows, Vector2 [] source
...
-				dpPtr = dpGch.AddrOfPinnedObject ();
-			}
-
-			var grid = GridWithColumns (cols, rows, spPtr, dpPtr);
-
-			spGch.Free ();
-			dpGch.Free ();
-
-			return grid;
+			fixed (Vector2* source_ptr = &sourcePositions [0])
```
Comment (spouliot) [inline src/SpriteKit/SKWarpGeometryGrid.cs]: same(s)

Comment (rolfbjarne) [inline src/SpriteKit/SKWarpGeometryGrid.cs]: Both sourcePositions and destPositions are guaranteed to have length > 0 due to argument validation earlier in the function.

File: src/AudioUnit/AudioUnit.cs (hunk 1 above)
Comment (rolfbjarne) [inline src/AudioUnit/AudioUnit.cs]: I can't take the address of an array:

    test.cs(7,25): error CS0208: Cannot take the address of, get the size of, or declare a pointer to a managed type `byte[]'

for this code:

    var bytes = new byte [0];
    fixed (byte* ptr = &bytes) {}

File: src/CoreFoundation/CFReadStream.cs (hunk 3 above)
Comment (rolfbjarne) [inline src/CoreFoundation/CFReadStream.cs]: buffer.Length can't be 0 here due to argument validation (count has to be >= 1, and offset + count has to be <= buffer.Length).

File: src/AudioUnit/AudioUnit.cs (hunk 1 above)
Comment (spouliot) [inline src/AudioUnit/AudioUnit.cs]: iirc it's `fixed (byte* ptr = bytes)`

Comment (rolfbjarne) [inline src/AudioUnit/AudioUnit.cs]: Yeah, that works.

**Pull request merged**